src/main/java/org/elasticsearch/index/translog/fs/FsTranslog.java

```diff
@@ -78,80 +81,120 @@ public void onRefreshSettings(Settings settings) {
+                      BigArrays bigArrays, Path[] locations) throws IOException {
         super(shardId, indexSettings);
+        ReadWriteLock rwl = new ReentrantReadWriteLock();
+        readLock = new ReleasableLock(rwl.readLock());
+        writeLock = new ReleasableLock(rwl.writeLock());
+
         this.indexSettingsService = indexSettingsService;
         this.bigArrays = bigArrays;
-        this.locations = indexStore.shardTranslogLocations(shardId);
+        this.locations = locations;
         for (Path location : locations) {
             Files.createDirectories(location);
         }
 
         this.type = FsTranslogFile.Type.fromString(indexSettings.get("index.translog.fs.type", FsTranslogFile.Type.BUFFERED.name()));
         this.bufferSize = (int) indexSettings.getAsBytesSize("index.translog.fs.buffer_size", ByteSizeValue.parseBytesSizeValue("64k")).bytes(); // Not really interesting, updated by IndexingMemoryController...
-        this.transientBufferSize = (int) indexSettings.getAsBytesSize("index.translog.fs.transient_buffer_size", ByteSizeValue.parseBytesSizeValue("8k")).bytes();
 
-        indexSettingsService.addListener(applySettings);
+        if (indexSettingsService != null) {
+            indexSettingsService.addListener(applySettings);
+        }
+        recoverFromFiles();
+        // new that we know which files are there, create a new current one.
+        current = createTranslogFile();
     }
 
-    public FsTranslog(ShardId shardId, @IndexSettings Settings indexSettings, Path location) throws IOException {
-        super(shardId, indexSettings);
-        this.indexSettingsService = null;
-        this.locations = new Path[]{location};
-        Files.createDirectories(location);
-        this.bigArrays = BigArrays.NON_RECYCLING_INSTANCE;
+    /** recover all translog files found on disk */
+    private void recoverFromFiles() throws IOException {
+        try (ReleasableLock _ = writeLock.acquire()) {
+            ArrayList<FsChannelImmutableReader> foundTranslogs = new ArrayList<>();
+            for (Path location : locations) {
+                try (DirectoryStream<Path> stream = Files.newDirectoryStream(location, TRANSLOG_FILE_PREFIX + "[0-9]*")) {
+                    for (Path file : stream) {
+                        final long id = parseIdFromFileName(file);
+                        if (id < 0) {
+                            // failed to extract id. logged in parseIdFromFileName
+                            continue;
+                        }
+                        idGenerator = Math.max(idGenerator, id + 1);
+                        final ChannelReference raf = new InternalChannelReference(id, location.resolve(getPath(id)), StandardOpenOption.READ);
+                        foundTranslogs.add(new FsChannelImmutableReader(id, raf, raf.channel().size(), FsChannelReader.UNKNOWN));
+                        logger.debug("found local translog with id [{}]", id);
```

Comment:
Not sure if this already includes the shard context, but if it doesn't can you change it to:

```
logger.debug("{} found local translog with id [{}]", this.shardId, id);
```